lib/Sema/TypeCheckDecl.cpp

```diff
@@ -2417,13 +2418,64 @@ NamingPatternRequest::evaluate(Evaluator &evaluator, VarDecl *VD) const {
   return namingPattern;
 }
 
-DeclRange
+namespace {
+
+// Utility class for deterministically ordering vtable entries for
+// synthesized methods.
+struct SortedFuncList {
+  using Entry = std::pair<std::string, AbstractFunctionDecl *>;
+  SmallVector<Entry, 2> elts;
+  bool sorted = false;
+
+  void add(AbstractFunctionDecl *afd) {
+    Mangle::ASTMangler mangler;
+    std::string mangledName;
+    if (auto *cd = dyn_cast<ConstructorDecl>(afd))
+      mangledName = mangler.mangleConstructorEntity(cd, /*allocator=*/false);
+    else if (auto *dd = dyn_cast<DestructorDecl>(afd))
+      mangledName = mangler.mangleDestructorEntity(dd, /*deallocating=*/false);
+    else
+      mangledName = mangler.mangleEntity(afd);
```

Comment:
Too bad we don't just have a member function on AbstractFuncDecl to compute this "mangledNameForSorting". I wonder if this functionality is duplicated elsewhere?

Reply:
mangleEntity() will assert if you pass in a destructor or constructor. The reason we have separate entry points for those is that usually they have multiple manglings, for the (de)allocating and (de)initializing entry points, respectively. I don't think it's worth factoring this out.